lib/ansible/modules/system/pbis_domain.py

```diff
@@ -0,0 +1,287 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {
+    "metadata_version": "1.1",
+    "status": ["preview"],
+    "supported_by": "community",
+}
+
+
+DOCUMENTATION = """
+---
+module: pbis_domain
+short_description: Join or leave AD with PBIS (open)
+description:
+  - This module allows to join or leave an Active Directory domain with PBIS (open).
+  - The module attempts to catch and fix a couple of common error scenario's when joining.
+version_added: "2.8"
+author:
+  - Luc Stroobant (@stroobl)
+requirements:
+  - PBIS (open) should be installed on the system.
+options:
+  name:
+    description:
+      - Domain name to join or leave.
+      - Not required to leave a domain.
+    type: str
+    aliases: [domain]
+  state:
+    description:
+      - Join or leave the domain?
+    type: str
+    default: present
+    choices: [absent, join, leave, present]
+  join_options:
+    description:
+      - List of options for the join command.
+      - Refer to the PBIS documentation for all possible options.
```

Comment:
We recommend that modules do not have an option like this, but instead have parameters to enable all kind of functionality that influences the command line options. The reason for this is that in the future we can play around with parameters (rename them, make dependencies, migrate from one option to another) which is not possible with people's individual option lists.

Another problem is that people will add options or do things that this was not designed for, and that results in support questions (i.e. the module not being idempotent as a result, or not behaving as expected). Which in the end is end-user support for the command, rather than the module.

We have experienced issues like this before, so this comes with quite some legacy :-)

Reply:
Fair point, changed that.